File: C/BSTree.c
#include <stdio.h>
#include <stdlib.h>

struct BSTree
{
    int data;
    struct BSTree *next;
    struct BSTree *prev;
};

struct BSTree* insert(struct BSTree* p,int val)
{
    struct BSTree* temp;
    if(p==NULL)
    {
        temp = (struct BSTree*)malloc(sizeof(struct BSTree));
        temp->data = val;
        temp->next=NULL;
        temp->prev=NULL;
        p=temp;
    }
    else if(val<p->data)
    {
        p->prev=insert(p->prev,val);
    }
    else
    {
        p->next=insert(p->next,val);
    }
    
    return p;
}

void inorder(struct BSTree* root)
{
    if(root!=NULL)
    {
        inorder(root->prev);
        printf("%d ",root->data);
        inorder(root->next);
    }
}

void preorder(struct BSTree* root)
{
    if(root!=NULL)
    {
        
        printf("%d ",root->data);
        preorder(root->prev);
        preorder(root->next);
        
    }
}

void postorder(struct BSTree* root)
{
    if(root!=NULL)
    {
        postorder(root->prev);
        postorder(root->next);
        printf("%d ",root->data);
        
    }
}


int main()
{
    struct BSTree* root = NULL;
    root = insert(root,15);
    root = insert(root,8);
    root = insert(root,18);
    root = insert(root,48);
    printf("Inorder: ");
    inorder(root);
    printf("\nPreorder: ");
    preorder(root);
    printf("\nPostorder: ");
    postorder(root);
}

File: Java/BSTree.java

 import java.util.*;
 
 class BSTree
 {

    static class Node
    {
        int data;
        Node left,right;
        Node(int val)
        {
            data=val;
            left=null;
            right=null;
        }
    }

    static Node insert(Node p,int val)
    {
        if(p==null)
        {
            Node temp = new Node(val);
            p=temp;
        }
        else if(val>p.data)
        {
            p.right=insert(p.right,val);
        }
        else
        {
            p.left=insert(p.left,val);
        }
        return p;
    }

    static void inorder(Node p)
    {
        if(p!=null)
        {
            inorder(p.left);
            System.out.print(p.data + " ");
            inorder(p.right);
        }
    }

    static void preorder(Node p)
    {
        if(p!=null)
        {
            System.out.print(p.data+" ");
            preorder(p.left);
            preorder(p.right);
        }
    }
    static void postorder(Node p)
    {
        if(p!=null)
        {
            postorder(p.left);
            postorder(p.right);
            System.out.print(p.data+" ");
        }
    } 

    public static void main(String args[])
    {
        Node root = null;
        root=insert(root,5);
        root=insert(root,3);
        root=insert(root,4);
        root=insert(root,1);
        root=insert(root,2);
        root=insert(root,6);
        System.out.print("Inorder: ");
        inorder(root);
        System.out.println();
        System.out.print("Preorder: ");
        preorder(root);
        System.out.println();
        System.out.print("Postorder: ");
        postorder(root);
        
    }


 }


